src/Tools/Source/CompilerGeneratorTools/Source/CSharpSyntaxGenerator/Grammar/GrammarGenerator.cs

```diff
@@ -0,0 +1,516 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Linq;
+using System.Reflection;
+using System.Text;
+using System.Text.RegularExpressions;
+using Microsoft.CodeAnalysis.CSharp;
+using static System.String;
+
+namespace CSharpSyntaxGenerator.Grammar
+{
+    internal class GrammarGenerator
+    {
+        private readonly ImmutableDictionary<string, TreeType> _nameToElement;
+
+        public GrammarGenerator(Tree tree)
+        {
+            // Syntax refers to a special pseudo-element 'Modifier'.  Just synthesize that since
+            // it's useful in the g4 grammar.
+            tree.Types.Add(new Node
+            {
+                Name = Modifier,
```

Comment:
I'm having trouble mentally binding this. What is `Modifier`, `SyntaxToken` and `Modifiers` referring to?

Reply:
literally those are just strings (but they appear in a few places).  So `Modifier = "Modifier"`.

Reply:
Maybe `ModifierName` would be preferred?  not sure.  but scroll down and you'll see at bottom.  can also move to the top if you think thta would help out.

Reply:
the basic idea though is that this is synthesizing a rule like `Modifier: 'abstract' | 'public' | ...` and later on other rules contain things like `Modifier*` in it.